### PR TITLE
Restful API Design Guidelines

### PR DESCRIPTION
**Pull request by johntopleyons**
Documentation from internal Confluence as implemented by RM/RAS microservices for BRES.

### CODE REVIEW

File: microservices/restful-api-design-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# RESTful API Design Guidelines
+
+## Introduction
+This page documents some guidelines to be used when designing and specifying RESTful APIs for microservices within the Survey Data Collection (SDC) platform.
+
+## HATEOAS
+We will not be building full [HATEOAS](https://en.wikipedia.org/wiki/HATEOAS)-complaint APIs as there is currently no clear requirement for the SDC platform that justifies the additional development complexity and effort.
+
+## URIs
+URIs should be all lowercase. The primary means for identifying a resource to access is via its UUID, as per the object identity strategy. The UUID should appear after the plural resource name without further qualification:
```
Comment (necrophonic) [inline microservices/restful-api-design-guidelines.md]: I agree with the lowercasing of the uris, but not sure about going into the detail of the uuid bit - that's quite RAS/RM specific at the moment?

File: microservices/restful-api-design-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+`GET /parties/ref/49900000001`
+
+Resource names in URIs should be pluralised, apart from where the identity of a secondary resource is used to further qualify the request, in which case the secondary resource name should be singular. For example, retrieving the collection exercises for a survey:
+
+`GET /collectionexercises/survey/cb0711c3-0ac8-41d3-ae0e-567e5a1ef87`
+
+## HTTP Status Codes
+Endpoints should return the appropriate HTTP status code. Including, but not limited to:
+
+* `HTTP 201 Created` - when a resource has been created
```
Comment (necrophonic) [inline microservices/restful-api-design-guidelines.md]: Not sure you need to list the codes - should just be able to state that correct codes should be returned as appropriate and potentially link to a resource list of standard HTTP codes

File: microservices/restful-api-design-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+* `HTTP 204 No Content` - when a request for a list of resources returns an empty list
+* `HTTP 400 Bad Request` - when invalid request parameters are supplied
+* `HTTP 404 Not Found` - when the requested resource could not be found
+
+## JSON Responses
+Primary resource identifiers within JSON responses should use an `id` attribute. Attributes that reference secondary resource types should be named with the resource type name in camel case with an `ID` suffix. External references i.e. identifiers that have a meaning to the business and are used outside the platform (e.g. reporting unit reference, case reference etc.) should be named with a `Ref` suffix:
```
Comment (necrophonic) [inline microservices/restful-api-design-guidelines.md]: Maybe bullet point for readability?

File: microservices/restful-api-design-guidelines.md (hunk 1 above)
Comment (johntopleyons) [reply microservices/restful-api-design-guidelines.md]: It is. I'd like to retain that detail, but I'll reword to make it clear the UUID stuff is as currently implemented by RAS/RM.

File: microservices/restful-api-design-guidelines.md (hunk 2 above)
Comment (johntopleyons) [reply microservices/restful-api-design-guidelines.md]: OK.

File: microservices/restful-api-design-guidelines.md (hunk 3 above)
Comment (johntopleyons) [reply microservices/restful-api-design-guidelines.md]: Good suggestion.

**Comment by NickGrantX** [comment]
Please resolve review comments, or can we close please?

**Comment by johntopleyons** [comment]
I am guessing that teams within SDC aren't referring to these when designing RESTful APIs, so I wish I hadn't wasted my time really.

**Pull request closed without merging**